build.assets/build-fido2.sh

```diff
@@ -0,0 +1,209 @@
+#!/bin/bash
+#
+# Builds libfido2 and dependencies, caching the resulting binaries in the local
+# filesystem.
+#
+# Run `build-fido2.sh build` to build libfido2 and its dependencies, at the
+# versions specified in the script.
+# Run `build-fido2.sh pkg_config_path` to print the path to the prior-built
+# libfido2-static.pc file.
+#
+# Written mainly for macOS builders.
+set -eu
+
+# Note: versions are the same as the corresponding git tags for each repo.
+readonly CBOR_VERSION=v0.9.0
+readonly CRYPTO_VERSION=OpenSSL_1_1_1o
+readonly FIDO2_VERSION=1.11.0
+
+readonly LIB_CACHE="$HOME/.teleport-fido2-cache"
+
+readonly PKGFILE_DIR="$LIB_CACHE/fido2-${FIDO2_VERSION}_cbor-${CBOR_VERSION}_crypto-${CRYPTO_VERSION}"
+
+fetch_and_build() {
+  local name="$1"      # eg, cbor
+  local version="$2"   # eg, v0.9.0
+  local url="$3"       # eg, https://github.com/...
+  local buildcmd="$4"  # eg, cbor_build, a bash function name
+  echo "$name: fetch and build" >&2
+
+  local tmp=''
+  tmp="$(mktemp -d)"
+  # Early expansion on purpose.
+  #shellcheck disable=SC2064
+  trap "rm -fr '$tmp'" exit
+
+  local fullname="$name-$version"
+  local install_path="$tmp/$fullname"
+
+  cd "$tmp"
+  git clone --depth=1 -b "$version" "$url"
+  cd "$(ls)"  # a single folder exists at this point
+  mkdir -p "$install_path"
+  eval "$buildcmd '$PWD' '$install_path'"
+
+  # Fix path in pkgconfig files.
+  local dest="$LIB_CACHE/$fullname"
+  find "$install_path" \
+    -name '*.pc' \
+    -exec sed -i '' "s@$install_path@$dest@g" {} +
+
+  # Check if another builder beat us. Builds _should_ be equivalent.
+  mkdir -p "$LIB_CACHE"
+  if [[ ! -d "$dest" ]]; then
+    echo "$name: moving $fullname to $dest" >&2
+    mv "$install_path" "$dest"
```

Comment:
Isn't this still racy - in the sense of erroring out if a race condition happens?

Reply:
Yep, there's a small change this will still fail due to a race condition. I don't think we have enough concurrent builds for it to be a serious problem, though.

Reply:
Adding a bit more detail, I chose to have it fail in the off-chance of a race instead of swallowing a possible mv error (say, `mv src dest || true`), which risks moving forward in a possible bad state. This should only happen if there is a race in the "fist batch" of building a new version, and retrying the failed builder would have it work (without having to rebuild anything, so not much time is lost). As said above, I think the chance for flakiness here is rare, so I didn't try to be clever about it.